pkg/applicationserver/applicationserver.go

```diff
@@ -540,6 +543,15 @@ func (as *ApplicationServer) handleUp(ctx context.Context, up *ttnpb.Application
 	}
 }
 
+func (as *ApplicationServer) handleSimulatedUp(ctx context.Context, up *ttnpb.ApplicationUp, link *link) error {
+	switch p := up.Up.(type) {
+	case *ttnpb.ApplicationUp_UplinkMessage:
```

Comment:
So can we or can we not simulate other upstream message types like in the original comment?

Reply:
We can. `handleSimulatedUp` does to simulated traffic what `handleUp` does to real traffic - process the contents of the `ApplicationUp`. For real traffic this means decrypt and decode for `UplinkMessage`, recalculating the queue on `DownlinkQueueInvalidated` and so forth. The simulated traffic does not need any special processing when it comes to other message types, so there are no switch cases for these.

Reply:
OK yeah I see now, thanks